californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java

```diff
@@ -556,34 +558,69 @@ public void receiveResponse(final Exchange exchange, final Response response) {
 					KeyUri key = getKey(exchange, exchange.getCurrentRequest());
 					clearBlock1Status(key);
 				default:
-					Response resp = new Response(response.getCode());
-					resp.setToken(exchange.getRequest().getToken());
-					if (exchange.getRequest().getType() == Type.CON) {
-						resp.setType(Type.ACK);
-						resp.setMID(exchange.getRequest().getMID());
+				}
+				// prepare the response as response to the original request
+				Response resp = new Response(response.getCode());
+				// adjust the token using the original request
+				resp.setToken(exchange.getRequest().getToken());
+				if (exchange.getRequest().getType() == Type.CON) {
+					resp.setType(Type.ACK);
+					// adjust MID also
+					resp.setMID(exchange.getRequest().getMID());
+				} else {
+					resp.setType(Type.NON);
+				}
+				resp.setSource(response.getSource());
+				resp.setSourcePort(response.getSourcePort());
+				resp.setPayload(response.getPayload());
+				resp.setOptions(response.getOptions());
+				exchange.setResponse(resp);
+				upper().receiveResponse(exchange, resp);
+				return;
+			}
+
+			KeyUri key = getKey(exchange, response);
+			BlockOption block = response.getOptions().getBlock2();
+			Block2BlockwiseStatus status = getBlock2Status(key);
+			if (status != null) {
+				// ongoing blockwise transfer
+				boolean starting = (block == null) || (block.getNum() == 0);
+				if (starting) {
+					if (status.isNew(response)) {
+						LOGGER.log(
+								Level.FINER,
+								"discarding outdated block2 transfer response for old notification {0}, current is {1}",
+								new Object[]{ status.getObserve(), response });
+						clearBlock2Status(key);
+						status.completeTransfer(exchange);
 					} else {
-						resp.setType(Type.NON);
+						LOGGER.log(
+								Level.FINER,
+								"discarding old notification [{0}] received during ongoing blockwise transfer: {1}",
```

Comment:
The log talk about "notification" but It seems to me that we can have block of "normal" response.